# Developer's Survey - PY 2021

In [1]:
# Importing Libs
import itertools
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

ModuleNotFoundError: No module named 'matplotlib'

In [ ]:
# Reseting figure size.
plt.rcParams['figure.figsize'] = [10, 5]

In [ ]:
# Loading data in to a dataframe.
df = pd.read_csv('enc_devs_py.csv')

## DataSet initial inspection.

In [107]:
# Printing the columns in the dataset.
for k,v in enumerate(df.columns):
    print(f'{k} - {v}')

0 - Marca temporal
1 - Modalidad
2 - Trabajo
3 - Tipo de trabajo
4 - DevOps
5 - Codificar por trabajo
6 - Experiencia en TI
7 - Experiencia en codificación 
8 - Primera línea de código en una máquina
9 - Educación
10 - Carreras o especialidades
11 - Importancia educación formal
12 - Género
13 - Edad actual 
14 - Salario mensual
15 - Programación, scripting y lenguajes de marcado
16 - Librerías o web frameworks
17 - Otras frameworks, librerías y herramientas
18 - Base de datos
19 - Plataformas
20 - Lenguaje favorito
21 - Lenguaje más odiado
22 - Framework favorito
23 - Framework más odiado
24 - Herramientas


 - We have 25 columns for this dataset. (At first sight we can see the names of each column are way to large. This needs to be fixed.)

In [108]:
df.shape

(342, 25)

- We have a df with ***342 rows*** and ***25 columns***.

## Data Dictionary
    - This data set has no "data dictionary" attached to it. So we have no way to know what each column means. For this reazon im going to make one with my own interpretation of the content in each column and its meaning.

- ***Modalidad***: The modality of the work that does the person whos answering the survey. Presencial / Remoto / Hibrido.
- ***Trabajo***: Origin of the work of the person whos answering the survey. Nacional o extranjeros.
- ***Tipo de trabajo***: Role of the work. Dev/Designer
- ***DevOps***: Role of the work in the DEVOPS field.
- ***Codificar por trabajo***: The work done requires "programming"?
- ***Experiencia en TI***: Years of experience in TI.
- ***Experiencia en codificación***: Years of experience in programming.
- ***Primera línea de código en una máquina***: It corresponds to the "age" that the respondent had when he wrote his first line of code on a machine.
- ***Educación***: Level of education.
- ***Carreras o especialidades***: Carreres or specializations.
- ***Importancia educación formal***: An estimation on the "importance of a formal education".
- ***Género***: Sex of the person who answers the survey.
- ***Edad actual***: Current age of the person who answer the survey.
- ***Salario mensual***: Monthly salary received.
- ***Programación, scripting y lenguajes de marcado***: Programming language's used.
- ***Librerías o web frameworks***: Frameworks or libraries used.
- ***Otras frameworks, librerías y herramientas***: Other frameworks or libraries.
- ***Base de datos***: Data Base used.
- ***Plataformas***: Working platforms.
- ***Lenguaje favorito***: Favorite language.
- ***Lenguaje más odiado***: Most hated language.
- ***Framework favorito***: Favorite framework.
- ***Framework más odiado***: Most hated framework.
- ***Herramientas***: Tools used to get the work done.

## Let's make some changes.

    - What i'm going to do here is to inspect each column and see if the content can be analized as it is. If not, i'll try to make'it "usable" making some changes to it. I'll be especially focused on the size of the content. Trying my best to avoid to work with long strings of text.

In [109]:
df.columns

Index(['Marca temporal', 'Modalidad', 'Trabajo', 'Tipo de trabajo', 'DevOps',
       'Codificar por trabajo', 'Experiencia en TI',
       'Experiencia en codificación ',
       'Primera línea de código en una máquina', 'Educación',
       'Carreras o especialidades', 'Importancia educación formal', 'Género',
       'Edad actual ', 'Salario mensual',
       'Programación, scripting y lenguajes de marcado',
       'Librerías o web frameworks',
       'Otras frameworks, librerías y herramientas', 'Base de datos',
       'Plataformas', 'Lenguaje favorito', 'Lenguaje más odiado',
       'Framework favorito', 'Framework más odiado', 'Herramientas'],
      dtype='object')

In [110]:
new_names = {
    'Marca temporal': 'marca_temp', 
    'Modalidad':'modalidad', 
    'Trabajo':'trabajo', 
    'Tipo de trabajo':'tipo_de_trabajo', 
    'DevOps':'devops',
    'Codificar por trabajo':'programar_x_trabajo', 
    'Experiencia en TI':'exp_en_IT', 
    'Experiencia en codificación ':'exp_en_programacion',
    'Primera línea de código en una máquina':'1_linea_de_codigo', 
    'Educación':'educacion',
    'Carreras o especialidades':'carr_especialidades', 
    'Importancia educación formal':'imp_ed_formal', 
    'Género':'genero',
    'Edad actual ':'edad_actual', 
    'Salario mensual':'salario_mensual',
    'Programación, scripting y lenguajes de marcado':'prog_script_leng',
    'Librerías o web frameworks':'lib_frameworks',
    'Otras frameworks, librerías y herramientas':'otras_lib_frameworks', 
    'Base de datos':'db',
    'Plataformas':'plataformas', 
    'Lenguaje favorito':'leng_fav', 
    'Lenguaje más odiado':'leng_mas_odiado',
    'Framework favorito':'fram_fav', 
    'Framework más odiado':'fram_mas_odiado', 
    'Herramientas':'herramientas'
}

# Renamin each column in the df.
df.rename(columns=new_names, inplace=True)


In [111]:
df.columns

Index(['marca_temp', 'modalidad', 'trabajo', 'tipo_de_trabajo', 'devops',
       'programar_x_trabajo', 'exp_en_IT', 'exp_en_programacion',
       '1_linea_de_codigo', 'educacion', 'carr_especialidades',
       'imp_ed_formal', 'genero', 'edad_actual', 'salario_mensual',
       'prog_script_leng', 'lib_frameworks', 'otras_lib_frameworks', 'db',
       'plataformas', 'leng_fav', 'leng_mas_odiado', 'fram_fav',
       'fram_mas_odiado', 'herramientas'],
      dtype='object')

- The new names are shorter and easy to write. None lost meaning in the process.

In [112]:
# Let's see the df head.
df.head()

marca_temp modalidad              trabajo  \
0  2021/03/01 5:56:11 p. m. GMT-3    Remoto  Nacional (Paraguay)   
1  2021/03/01 5:57:32 p. m. GMT-3   Hibrído  Nacional (Paraguay)   
2  2021/03/01 5:57:57 p. m. GMT-3   Hibrído  Nacional (Paraguay)   
3  2021/03/01 5:59:04 p. m. GMT-3   Hibrído                Ambas   
4  2021/03/01 5:59:44 p. m. GMT-3    Remoto  Nacional (Paraguay)   

                                     tipo_de_trabajo  \
0                                       QA o testing   
1  Desarrollador full-stack;Aplicaciones de escri...   
2                            Desarrollador front-end   
3                           Desarrollador full-stack   
4                               Desarrollador mobile   

                               devops programar_x_trabajo        exp_en_IT  \
0                            Educador                  No  Menos de 5 años   
1  Administrador de sistemas;Educador                  Sí     15 a 19 años   
2                                 NaN                  Sí  Menos de 5 años   
3           Administrador de sistemas                  Sí       5 a 9 años   
4                                 NaN                  Sí       5 a 9 años   

  exp_en_programacion 1_linea_de_codigo  \
0                 NaN      22 a 23 años   
1        15 a 19 años      20 a 21 años   
2     Menos de 5 años      20 a 21 años   
3          5 a 9 años      10 a 11 años   
4          5 a 9 años      16 a 17 años   

                                           educacion  ...  \
0                            Educación universitaria  ...   
1                Educación universitaria;Autodidacta  ...   
2                            Educación universitaria  ...   
3         Educación universitaria;Master;Autodidacta  ...   
4  Educación básica;Educación secundaria;Autodidacta  ...   

                              prog_script_leng  \
0                      JavaScript;Java;C++;SQL   
1                   JavaScript;C#;HTML/CSS;SQL   
2  JavaScript;TypeScript;PHP;C;C#;HTML/CSS;SQL   
3  JavaScript;TypeScript;PHP;HTML/CSS;SQL;Dart   
4                        PHP;Java;Kotlin;Swift   

                         lib_frameworks otras_lib_frameworks  \
0                                   NaN                  NaN   
1                  ASP.NET;ASP.NET Core         .NET;Xamarin   
2  React.js;Angular;JQuery;ASP.NET Core    Opció.NET Coren 3   
3     Angular;Angular.js;JQuery;Laravel              Flutter   
4                               Laravel                  NaN   

                                                  db  \
0                           MySQL;PostgreSQL;IBM DB2   
1                         MySQL;Microsoft SQL Server   
2                                                NaN   
3  MySQL;MariaDB;PostgreSQL;Microsoft SQL Server;...   
4                                              MySQL   

                                         plataformas leng_fav leng_mas_odiado  \
0                  GNU/Linux;Microsoft Windows;MacOS      SQL             NaN   
1  GNU/Linux;Microsoft Windows;Microsoft Azure;Wo...       C#            Java   
2                  GNU/Linux;Microsoft Windows;MacOS       C#         Genexus   
3      GNU/Linux;Microsoft Windows;MacOS;Android;iOS      PHP             NaN   
4                            Microsoft Windows;MacOS   Kotlin             C++   

       fram_fav fram_mas_odiado  \
0           NaN             NaN   
1  ASP.NET Core         Laravel   
2  ASP.NET Core          Django   
3       Laravel             NaN   
4       Laravel             NaN   

                                        herramientas  
0           GitHub;Slack;Jira;Microsoft Team;Discord  
1              GitHub;Microsoft Team;Microsoft Azure  
2  GitHub;Slack;GitLab;Google Suite (Docs, Meet, ...  
3                                GitHub;Slack;Trello  
4             GitHub;Slack;Confluence;Trello;Discord  

[5 rows x 25 columns]

### Modalidad

In [113]:
# Let's see the content of the column 'modalidad'.
df['modalidad'].value_counts().keys()

Index(['Remoto', 'Hibrído', 'Presencial', 'Ninguna de las anteriores'], dtype='object')

In [114]:
# One of the options 'Ninguna de las anteriores' its a long string. Let's make it shorter.

In [115]:
df['modalidad'].replace({
    'Ninguna de las anteriores': 'Ninguna',
}, regex=True, inplace=True)

In [116]:
df.modalidad.value_counts()

Remoto        145
Hibrído       110
Presencial     78
Ninguna         9
Name: modalidad, dtype: int64

### Trabajo

In [117]:
# First let's see the content of the column.
df['trabajo'].value_counts()

Nacional (Paraguay)    259
Ambas                   44
Internacional           39
Name: trabajo, dtype: int64

In [118]:
# No need of changes here.

### Tipo de Trabajo

In [119]:
# Column content
df['tipo_de_trabajo'].value_counts()

Desarrollador full-stack                                                                                                                      55
Desarrollador back-end                                                                                                                        29
Desarrollador front-end                                                                                                                       16
Aplicaciones de escritorio o empresariales                                                                                                    15
Desarrollador back-end;Desarrollador front-end                                                                                                13
                                                                                                                                              ..
Desarrollador full-stack;Desarrollador mobile;Administrador de base de datos;Project Leader                                       

In [120]:
# Here we see a series of composed strings. So wee need to do something to make it shorter.

In [121]:
df['tipo_de_trabajo'].replace({
    'Desarrollador':'Dev',
    'Project Leader':'Project Lead',
    'Aplicaciones':'Aps',
    'Administrador de base de datos':'DB Admin',
    'de escritorio o empresariales':'para Esc. o Emp.',
    'sistemas embebidos':'sist. embebidos',
    
    
},regex=True, inplace=True)

In [244]:
set(itertools.chain.from_iterable([i.split(';') for i in df['tipo_de_trabajo'].value_counts().keys()]))
    

{'Aps para Esc. o Emp.',
 'DB Admin',
 'Dev back-end',
 'Dev de sist. embebidos',
 'Dev de videojuegos',
 'Dev front-end',
 'Dev full-stack',
 'Dev mobile',
 'Project Lead',
 'QA o testing'}

In [245]:
# Now it is possible to work with the content of this column.

### DevOps

In [251]:
# Show the content of the column.
set(itertools.chain.from_iterable([i.split(';') for i in df.devops.value_counts().keys()]))
    

{'Data Scientist ML Specialist',
 'Diseñador',
 'Educador',
 'Inv. Cientifico',
 'Sys Admin'}

In [252]:
df['devops'].replace({
    'Científico de datos o especialista Machine Learning':'Data Scientist ML Specialist',
    'Administrador de sistemas':'Sys Admin',
    'Investigador científico':'Inv. Cientifico',
}, regex=True, inplace = True)

In [255]:
set(itertools.chain.from_iterable([i.split(';') for i in df['devops'].value_counts().keys()]))
    

{'Data Scientist ML Specialist',
 'Diseñador',
 'Educador',
 'Inv. Cientifico',
 'Sys Admin'}

In [256]:
# Now it is possible to work with the content of this column.

### 'programar_x_trabajo' 

In [257]:
# Show the column content.
df['programar_x_trabajo'].value_counts()

Sí    317
No     25
Name: programar_x_trabajo, dtype: int64

In [258]:
# No need to do changes.

### 'exp_en_IT'

In [259]:
df['exp_en_IT'].value_counts().keys()

Index(['5-9', '<5', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39',
       '>50', '45-49'],
      dtype='object')

In [260]:
# Numbers and strings together. I am going to convert it into only numbers but of type string. 
# All measurements are in "years".

In [261]:
df['exp_en_IT'].replace({
    "5 a 9 años":"5-9",
    "Menos de 5 años":"<5",
    "10 a 14 años":"10-14",
    "15 a 19 años":"15-19",
    "20 a 24 años":"20-24",
    "25 a 29 años":"25-29",
    "30 a 34 años":"30-34",
    "35 a 39 años":"35-39",
    "más de 50 años":">50",
    "45 a 49 años":"45-49",
},regex=True, inplace =True)

In [262]:
# Show the changes.
df['exp_en_IT'].value_counts()

5-9      124
<5        94
10-14     82
15-19     20
20-24      9
25-29      5
30-34      4
35-39      2
>50        1
45-49      1
Name: exp_en_IT, dtype: int64

### 'exp_en_programacion'

In [263]:
# Show the column content.
df['exp_en_programacion'].value_counts().keys()

Index(['5-9', '<5', '10-14', '15-19', '20-24', '25-29', '30-34', '35-39'], dtype='object')

In [264]:
# The same as the previous case. Numbers and strings together. 
# I am going to convert it into only numbers but of type string. All measurements are in "years".

In [265]:
df['exp_en_programacion'].replace({
    "5 a 9 años":"5-9",
    "Menos de 5 años":"<5",
    "10 a 14 años":"10-14",
    "15 a 19 años":"15-19",
    "20 a 24 años":"20-24",
    "25 a 29 años":"25-29",
    "30 a 34 años":"30-34",
    "35 a 39 años":"35-39",
}, regex=True, inplace=True)

In [266]:
# Show the changes.
df['exp_en_programacion'].value_counts()

5-9      125
<5       111
10-14     64
15-19     20
20-24      7
25-29      6
30-34      4
35-39      2
Name: exp_en_programacion, dtype: int64

### '1_linea_de_codigo' 

In [267]:
# Printing the values in the column.
df['1_linea_de_codigo'].value_counts().keys()

Index(['<10', '10-11', '14-15', '12-13', '16-17', '18-19', '20-21', '>30',
       '26-27', '22-23', '28-29', '24-25'],
      dtype='object')

In [268]:
# The same as the previous case. Numbers and strings together. 
# I am going to convert it into only numbers but of type string. All measurements are in "years".

In [269]:
df['1_linea_de_codigo'].replace({
    'Menos de 10 años': '<10',
    '10 a 11 años':'10-11', 
    '14 a 15 años':'14-15', 
    '12 a 13 años':'12-13',
    '16 a 17 años':'16-17', 
    '18 a 19 años':'18-19', 
    '20 a 21 años':'20-21', 
    'más de 30 años':'>30',
    '26 a 27 años':'26-27', 
    '22 a 23 años':'22-23', 
    '24 a 25 años':'24-25', 
    '28 a 29 años':'28-29',
},regex=True, inplace=True)

In [270]:
# Printing the changes.
df['1_linea_de_codigo'].value_counts()

<10      101
10-11     55
14-15     51
12-13     37
16-17     31
18-19     27
20-21     16
>30        6
26-27      4
22-23      4
28-29      3
24-25      3
Name: 1_linea_de_codigo, dtype: int64

### 'educación'

In [276]:
set(itertools.chain.from_iterable([i.split(';') for i in df['educacion'].value_counts().keys()]))

{'Autodidacta',
 'Doctorado',
 'Ed. Basica',
 'Ed. Secundaria',
 'Ed. Universitaria',
 'Master'}

In [277]:
# Some of the strings in the column can be shortened.

In [278]:
df['educacion'].replace({
    'Educación universitaria':'Ed. Universitaria',
    'Educación básica': 'Ed. Basica',
    'Educación secundaria': 'Ed. Secundaria',
    'Autodidacta': 'Autodidacta',
},regex=True, inplace=True)

In [279]:
# Show the changes made.
df['educacion'].value_counts()

Ed. Universitaria                                                           151
Ed. Basica;Ed. Secundaria;Ed. Universitaria                                  42
Ed. Basica;Ed. Secundaria;Ed. Universitaria;Autodidacta                      36
Ed. Universitaria;Autodidacta                                                29
Master                                                                       17
Ed. Secundaria                                                               12
Ed. Basica;Ed. Secundaria                                                    11
Ed. Basica;Ed. Secundaria;Autodidacta                                        10
Ed. Secundaria;Autodidacta                                                    8
Ed. Universitaria;Master                                                      8
Ed. Basica;Ed. Secundaria;Ed. Universitaria;Master                            4
Ed. Universitaria;Master;Autodidacta                                          3
Autodidacta                             

### 'carr_especialidades' 

In [280]:
df['carr_especialidades'].value_counts().keys()

Index(['Ing. Informática', 'Análisis de Sist.',
       'Programación;Análisis de Sist.', 'Programación;Ing. Informática',
       'Programación', 'Programación;Análisis de Sist.;Ing. Informática',
       'Ninguna', 'Análisis de Sist.;Ing. Informática', 'Ing. Electrónica',
       'Otras', 'Programación;Análisis de Sist.;Otras',
       'Ing. Informática;Otras', 'Otras Ing.', 'Análisis de Sist.;Otras',
       'Programación;Análisis de Sist.;Diseño Gráfico',
       'Programación;Diseño Gráfico',
       'Programación;Análisis de Sist.;Ing. Informática;Otras',
       'Programación;Ing. Informática;Otras',
       'Programación;Diseño Gráfico;Otras', 'Programación;Otras',
       'Análisis de Sist.;Diseño Gráfico', 'Ing. Informática;Diseño Gráfico',
       'Programación;Análisis de Sist.;Ing. Informática;Matemático',
       'Programación;Otras Ing.;Diseño Gráfico', 'Matemático',
       'Programación;Análisis de Sist.;Ing. Civil;Otras Ing.',
       'Análisis de Sist.;Ing. Electrónica',
       'Pr

In [281]:
# Again, some of the values in this column are long strings. I'm going to make it a bit shorter.

In [283]:
set(itertools.chain.from_iterable([i.split(';') for i in df['carr_especialidades'].value_counts().keys()]))

{'Análisis de Sist.',
 'Diseño Gráfico',
 'Ing. Civil',
 'Ing. Electrónica',
 'Ing. Informática',
 'Matemático',
 'Ninguna',
 'Otras',
 'Otras Ing.',
 'Programación'}

In [284]:
# Replacing careers or specialities for shorter strings names.
df['carr_especialidades'].replace({
    'Ingeniero Informática': 'Ing. Informática',
    'Análisis de sistemas':'Análisis de Sist.',
    'Ingeniero Electrónica':'Ing. Electrónica',
    'Otras ingenerías':'Otras Ing.',
    'Diseño gráfico':'Diseño Gráfico',
    'Ingeniero Civil':'Ing. Civil',
},regex=True, inplace=True)

In [285]:
# Printing the changes.
df['carr_especialidades'].value_counts()

Ing. Informática                                                        107
Análisis de Sist.                                                        69
Programación;Análisis de Sist.                                           33
Programación;Ing. Informática                                            21
Programación                                                             17
Programación;Análisis de Sist.;Ing. Informática                          17
Ninguna                                                                  15
Análisis de Sist.;Ing. Informática                                       11
Ing. Electrónica                                                          6
Otras                                                                     5
Programación;Análisis de Sist.;Otras                                      5
Ing. Informática;Otras                                                    4
Otras Ing.                                                                4
Análisis de 

### 'imp_ed_formal' 

In [286]:
# Show the column content.
df['imp_ed_formal' ].value_counts().keys()

Index(['+ || -', 'BASTANTE', 'MUY', 'NO', 'CRITICO'], dtype='object')

In [287]:
# Let's make some changes to get shorter strings.

df['imp_ed_formal' ].replace({
    'Más o menos importante': '+ || -',
    'Bastante importante':'BASTANTE', 
    'Muy importante':'MUY',
    'No importa, no es necesario':'NO', 
    'Crítico':'CRITICO',
},regex=True, inplace=True)

In [288]:
# Print the changes made.
df['imp_ed_formal' ].value_counts().keys()

Index(['+ || -', 'BASTANTE', 'MUY', 'NO', 'CRITICO'], dtype='object')

### 'genero'

In [289]:
# Show the column content.
df['genero'].value_counts().keys()

Index(['HOMBRE', 'MUJER', 'NO COMPARTO'], dtype='object')

In [290]:
# GROUP GENRES BY MINORITIES.
# I know in advance that the categories 'Prefiero no compartir', 'Otros' and 'Género' have 
# very low values and do not represent a relevant number.
# For this reason I am going to group them in a single category 'I DO NOT SHARE'.

In [291]:
# Making the changes. 
df['genero'].replace({
    'Hombre':'HOMBRE', 
    'Mujer':'MUJER',
    'Prefiero no compartir':'NO COMPARTO', 
    'Otros':'NO COMPARTO', 
    'Género':'NO COMPARTO',
}, regex=True, inplace=True)

In [292]:
# Printing the changes.
df['genero'].value_counts()

HOMBRE         279
MUJER           49
NO COMPARTO      6
Name: genero, dtype: int64

In [293]:
df.columns

Index(['marca_temp', 'modalidad', 'trabajo', 'tipo_de_trabajo', 'devops',
       'programar_x_trabajo', 'exp_en_IT', 'exp_en_programacion',
       '1_linea_de_codigo', 'educacion', 'carr_especialidades',
       'imp_ed_formal', 'genero', 'edad_actual', 'salario_mensual',
       'prog_script_leng', 'lib_frameworks', 'otras_lib_frameworks', 'db',
       'plataformas', 'leng_fav', 'leng_mas_odiado', 'fram_fav',
       'fram_mas_odiado', 'herramientas'],
      dtype='object')

### 'edad_actual' 

In [294]:
# Print the column content.
df['edad_actual'].value_counts().keys()

Index(['30-34', '25-29', '35-39', '20-24', '40-44', '15-19', '45-49', '55-59',
       '50-54'],
      dtype='object')

NOTE: The content of this column could be more useful if it were single data not ranges of data.

In [295]:
# The same thing happens as in the previous cases. 
# Time series in the form of numbers and letters. Let's make some changes.

In [296]:
# Make some changes by replacing.
df['edad_actual' ].replace({
    '30 a 34 años': '30-34', 
    '25 a 29 años':'25-29', 
    '35 a 39 años':'35-39', 
    '20 a 24 años':'20-24', 
    '40 a 44 años':'40-44',  
    '15 a 19 años':'15-19',  
    '50 a 54 años':'50-54',  
    '55 a 59 años':'55-59', 
    '45 a 49 años':'45-49', 
}, regex=True, inplace=True)

In [297]:
# Show the changes.
df['edad_actual' ].value_counts()

30-34    120
25-29     98
35-39     51
20-24     42
40-44     16
15-19      6
45-49      3
55-59      3
50-54      3
Name: edad_actual, dtype: int64

### 'salario_mensual'

In [298]:
# First lets see what's in the column.
df['salario_mensual'].value_counts().keys()

Index(['5 a 7 M. Gs.', '7 a 9 M. Gs.', '3.5 a 5 M. Gs.', '9 a 12 M. Gs.',
       'MINIMO a 3.5 M. Gs.', '12 a 15 M. Gs.', '> 27 M. Gs.',
       '15 a 18 M. Gs.', '< MINIMO Gs.', '18 a 21 M. Gs.', 'MINIMO',
       '21 a 24 M. Gs.', '24 a 27 M. Gs.'],
      dtype='object')

In [299]:
# The column content it's a series of long strings. I'm going to change that.

In [300]:
df["salario_mensual"].replace({
    "Entre 5.000.000 a 7.000.000 gs":"5 a 7 M. Gs.",
    "Entre 7.000.000 a 9.000.000 gs":"7 a 9 M. Gs.",
    "Entre 3.500.000 a 5.000.000 gs":"3.5 a 5 M. Gs.",
    "Entre 9.000.000 a 12.000.000 gs":"9 a 12 M. Gs.",
    "Entre 2.192.839 gs a 3.500.000 gs":"MINIMO a 3.5 M. Gs.",
    "Entre 12.000.000 a 15.000.000 gs":"12 a 15 M. Gs.",
    "Más de 27.000.000 gs":"> 27 M. Gs.",
    "Entre 15.000.000 a 18.000.000 gs":"15 a 18 M. Gs.",
    "Menos del salario mínimo ( 2.192.839 gs)":"< MINIMO Gs.",
    "Entre 18.000.000 a 21.000.000 gs":"18 a 21 M. Gs.",
    "Salario mínimo 2.192.839 gs":"MINIMO",
    "Entre 21.000.000 a 24.000.000 gs":"21 a 24 M. Gs.",
    "Entre 24.000.000 a 27.000.000 gs":"24 a 27 M. Gs.",
}, inplace=True)

In [301]:
# Print the changes. 
df['salario_mensual'].value_counts()

5 a 7 M. Gs.           62
7 a 9 M. Gs.           50
3.5 a 5 M. Gs.         46
9 a 12 M. Gs.          45
MINIMO a 3.5 M. Gs.    39
12 a 15 M. Gs.         28
> 27 M. Gs.            17
15 a 18 M. Gs.         14
< MINIMO Gs.           13
18 a 21 M. Gs.          9
MINIMO                  8
21 a 24 M. Gs.          7
24 a 27 M. Gs.          4
Name: salario_mensual, dtype: int64

### 'prog_script_leng'

In [302]:
df['prog_script_leng'].value_counts()

JavaScript;PHP;HTML/CSS;SQL                                                 10
JavaScript;TypeScript;Python;Java;HTML/CSS;SQL                               5
JavaScript;TypeScript;Java;HTML/CSS;SQL                                      5
JavaScript;PHP;HTML/CSS;SQL;Bash/Shell/PowerShell                            5
JavaScript;TypeScript;PHP;Python;Java;HTML/CSS;SQL;Bash/Shell/PowerShell     4
                                                                            ..
Python;C#;VBA                                                                1
JavaScript;Python;Java;C;HTML/CSS;Genexus                                    1
JavaScript;PHP;C#;HTML/CSS                                                   1
JavaScript;PHP;C;HTML/CSS;Fox Pro                                            1
JavaScript;Python;Java;C;HTML/CSS;SQL;Genexus                                1
Name: prog_script_leng, Length: 270, dtype: int64

In [303]:
set(list(itertools.chain.from_iterable([i.split(';') for i in df['prog_script_leng'].value_counts().keys()])))

{'Assembly',
 'Bash/Shell/PowerShell',
 'C',
 'C#',
 'C++',
 'COBOL',
 'Dart',
 'Fox Pro',
 'Genexus',
 'Go',
 'HTML/CSS',
 'Haskell',
 'Java',
 'JavaScript',
 'Kotlin',
 'Objective-C',
 'Oracle form',
 'PHP',
 'Perl',
 'Python',
 'R',
 'RPG',
 'Ruby',
 'Rust',
 'SQL',
 'Scala',
 'Swift',
 'TypeScript',
 'VBA'}

In [304]:
# This are short strings so we can work with this.

### 'lib_frameworks'

In [305]:
# Printing the column content.
set(list(itertools.chain.from_iterable([i.split(';') for i in df['lib_frameworks'].value_counts().keys()])))

{'ASP.NET',
 'ASP.NET Core',
 'Angular',
 'Angular.js',
 'CakePHP',
 'CodeIgniter',
 'Django',
 'Express',
 'Flask',
 'Gatsby',
 'JQuery',
 'Laravel',
 'React.js',
 'Ruby on Rails',
 'Spring',
 'Symfony',
 'Vue.js'}

### 'otras_lib_frameworks' 

In [306]:
# Print the column content
df['otras_lib_frameworks' ].value_counts().keys()

Index(['Node.js', '.NET', '.NET;NET Core', 'Node.js;React Native',
       'Node.js;Flutter', 'Node.js;.NET', 'Flutter', 'Node.js;Cordova',
       'Node.js;.NET;NET Core', 'React Native', 'Cordova',
       'Node.js;.NET;Flutter', 'Node.js;Pandas', 'Ansible',
       'Node.js;React Native;Unity 3D;Flutter',
       'Node.js;Pandas;TensorFlow;React Native;Keras;Torch/PyTorch',
       'Pandas;TensorFlow;Keras', 'Node.js;React Native;Cordova', 'Pandas',
       'Node.js;React Native;Flutter', 'Flutter;Xamarin', '.NET;Xamarin',
       'Node.js;Pandas;TensorFlow', 'Node.js;.NET;React Native',
       'Node.js;Keras;Hadoop', 'Node.js;Pandas;Cordova',
       'Node.js;Pandas;Unity 3D', 'Node.js;.NET;Unity 3D;Unreal Engine',
       'Node.js;TensorFlow;Cordova',
       'Pandas;TensorFlow;Unity 3D;Flutter;Keras;Chef',
       '.NET;NET Core;Pandas;Cordova', 'React Native;Unity 3D;Flutter',
       '.NET;Pandas;TensorFlow;Ansible;Keras;Torch/PyTorch;Hadoop;Puppet;Chef',
       'Node.js;.NET;NET Core;React

In [307]:
# Let's see if some name needs to be changed.
set(list(itertools.chain.from_iterable([i.split(';') for i in df['otras_lib_frameworks'].value_counts().keys()])))

{'.NET',
 'Ansible',
 'Chef',
 'Cordova',
 'Flutter',
 'Hadoop',
 'Keras',
 'NET Core',
 'Node.js',
 'Opció.NET Coren 3',
 'Pandas',
 'Puppet',
 'React Native',
 'TensorFlow',
 'Torch/PyTorch',
 'Unity 3D',
 'Unreal Engine',
 'Xamarin'}

### 'db'

In [308]:
# Print the column values.
df['db'].value_counts().keys()

Index(['PostgreSQL', 'MySQL;PostgreSQL;Oracle',
       'MySQL;PostgreSQL;Microsoft SQL Server', 'MySQL;PostgreSQL',
       'PostgreSQL;Oracle', 'MySQL;MariaDB', 'Microsoft SQL Server',
       'MySQL;MariaDB;PostgreSQL',
       'MySQL;PostgreSQL;Microsoft SQL Server;Oracle', 'Oracle',
       ...
       'MySQL;MariaDB;PostgreSQL;Microsoft SQL Server;SQLite;IBM DB2',
       'MySQL;MariaDB;PostgreSQL;MongoDB;Redis',
       'MySQL;MariaDB;MongoDB;Firebase',
       'MySQL;MariaDB;PostgreSQL;SQLite;MongoDB;Firebase',
       'MySQL;MariaDB;PostgreSQL;SQLite;MongoDB;Redis;Couchbase',
       'Microsoft SQL Server;SQLite;Oracle',
       'MySQL;MariaDB;PostgreSQL;Microsoft SQL Server;MongoDB;Redis;Elasticsearch',
       'MySQL;PostgreSQL;Microsoft SQL Server;SQLite;MongoDB;Oracle;Couchbase;Elasticsearch',
       'MySQL;MariaDB;PostgreSQL;Oracle;Elasticsearch',
       'MySQL;MariaDB;PostgreSQL;Microsoft SQL Server;MongoDB'],
      dtype='object', length=154)

In [309]:
set(itertools.chain.from_iterable([i.split(';') for i in df['db'].value_counts().keys()]))

{'Cassandra',
 'Couchbase',
 'DynamoDB',
 'Elasticsearch',
 'Firebase',
 'IBM DB2',
 'MariaDB',
 'Microsoft SQL Server',
 'MongoDB',
 'MySQL',
 'Oracle',
 'PostgreSQL',
 'Redis',
 'SQLite'}

In [310]:
# All looks fine.

### 'plataformas' 

In [311]:
# Print the column content.
df['plataformas' ].value_counts().keys()

Index(['GNU/Linux;Microsoft Windows', 'Microsoft Windows', 'GNU/Linux',
       'GNU/Linux;Microsoft Windows;Docker',
       'GNU/Linux;Microsoft Windows;Android',
       'GNU/Linux;Microsoft Windows;MacOS', 'GNU/Linux;MacOS',
       'Microsoft Windows;Android', 'GNU/Linux;Microsoft Windows;Docker;AWS',
       'GNU/Linux;Microsoft Windows;AWS',
       ...
       'GNU/Linux;Microsoft Windows;Docker;Arduino',
       'GNU/Linux;Microsoft Windows;Docker;Kubernetes;Heroku',
       'GNU/Linux;Microsoft Windows;MacOS;Docker;AWS;Heroku',
       'GNU/Linux;Microsoft Windows;Docker;AWS;Android;Raspberry Pi;Kubernetes',
       'GNU/Linux;Microsoft Windows;MacOS;Docker;AWS;Android;Raspberry Pi;WordPress;Heroku;Arduino',
       'GNU/Linux;MacOS;Docker;AWS;Raspberry Pi;Microsoft Azure;WordPress;Google Cloud Platform;Kubernetes;Heroku',
       'GNU/Linux;Microsoft Windows;MacOS;AWS;Android;iOS',
       'GNU/Linux;Microsoft Windows;MacOS;Docker;iOS;Heroku',
       'GNU/Linux;Microsoft Windows;MacOS;AWS

In [312]:
set(itertools.chain.from_iterable([i.split(';') for i in df['plataformas'].value_counts().keys()]))

{'AWS',
 'Android',
 'Arduino',
 'Docker',
 'GNU/Linux',
 'Google Cloud Platform',
 'Heroku',
 'Kubernetes',
 'MacOS',
 'Microsoft Azure',
 'Microsoft Windows',
 'Raspberry Pi',
 'WordPress',
 'iOS'}

In [313]:
# Everything looks good here.

### 'leng_fav' 

In [203]:
# Printing the column values.
df['leng_fav' ].value_counts().keys()

Index(['Java', 'Python', 'JavaScript', 'PHP', 'C#', 'SQL', 'TypeScript',
       'HTML/CSS', 'Oracle form', 'Kotlin', 'Genexus', 'Go', 'C++', 'Dart',
       'C', 'Ruby', 'Otro', 'VBA', 'Swift', 'Bash/Shell/PowerShell', 'Rust',
       'COBOL', 'OpcióFox Pron 27'],
      dtype='object')

In [204]:
# Everything its fine here.

### 'leng_mas_odiado'

In [206]:
# Printing the column values.
df['leng_mas_odiado'].value_counts().keys()

Index(['Java', 'PHP', 'JavaScript', 'Fox Pro', 'Genexus', 'C', 'C++',
       'Assembly', 'VBA', 'COBOL', 'HTML/CSS', 'Ruby', 'Oracle form',
       'Bash/Shell/PowerShell', 'Objective-C', 'SQL', 'RPG', 'C#', 'Spring',
       'Perl', 'TypeScript', 'R', 'Haskell', 'Swift', 'Go', 'Kotlin',
       'Python'],
      dtype='object')

### 'fram_fav' 

In [213]:
# Print the column values.
df['fram_fav' ].value_counts().keys()

Index(['React.js', 'Laravel', 'Spring', 'Django', 'Angular', 'Vue.js',
       'ASP.NET', 'Express', 'ASP.NET Core', 'JQuery', 'Ruby on Rails',
       'Angular.js', 'Flask', 'CakePHP', 'CodeIgniter'],
      dtype='object')

In [214]:
# Looks good.

### 'fram_mas_odiado' 

In [215]:
# Print the column values.
df['fram_mas_odiado' ].value_counts().keys()

Index(['Spring', 'JQuery', 'ASP.NET', 'Angular', 'Laravel', 'Angular.js',
       'React.js', 'Ruby on Rails', 'CakePHP', 'ASP.NET Core', 'CodeIgniter',
       'Django', 'Express', 'Vue.js', 'Flask', 'Gastby'],
      dtype='object')

In [216]:
# No problem here.

### 'herramientas'

In [221]:
# Print column content.
df['herramientas'].value_counts().keys()

Index(['GitHub;GitLab',
       'GitHub;Slack;GitLab;Google Suite (Docs, Meet, etc);Trello',
       'GitHub;Slack;GitLab;Jira;Google Suite (Docs, Meet, etc);Confluence;Microsoft Team;Trello',
       'GitHub;Slack;GitLab;Jira;Google Suite (Docs, Meet, etc);Confluence',
       'GitHub;Slack;GitLab;Google Suite (Docs, Meet, etc);Trello;Discord',
       'Microsoft Team', 'GitHub;GitLab;Google Suite (Docs, Meet, etc);Trello',
       'GitHub;Slack;GitLab;Jira;Google Suite (Docs, Meet, etc);Microsoft Team;Trello',
       'GitHub;Slack;GitLab;Jira;Google Suite (Docs, Meet, etc);Trello',
       'GitHub;GitLab;Google Suite (Docs, Meet, etc)',
       ...
       'Slack;GitLab;Trello',
       'GitHub;Slack;GitLab;Jira;Google Suite (Docs, Meet, etc);Confluence;Microsoft Team;Trello;Facebook Workplace;Discord',
       'GitHub;GitLab;Google Suite (Docs, Meet, etc);Confluence;Microsoft Team;Microsoft Azure',
       'Slack;Google Suite (Docs, Meet, etc)',
       'GitHub;GitLab;Jira;Google Suite (Docs, Me

In [229]:
set(itertools.chain.from_iterable([i.split(';') for i in df['herramientas'].value_counts().keys()]))

{'Confluence',
 'Discord',
 'Facebook Workplace',
 'GitHub',
 'GitLab',
 'Google Suite',
 'Jira',
 'Microsoft Azure',
 'Microsoft Team',
 'Slack',
 'Trello'}

In [230]:
# "Google Suite" it's a large string. So im gona make it shorter.

In [231]:
tools = []
for i in df['herramientas'].value_counts().keys():
    for j in i.split(';'):
        tools.append(j)

In [232]:
def replace_google(row):
    if('Google Suite (Docs, Meet, etc)' in str(row)):
        return(row.replace('Google Suite (Docs, Meet, etc)', 'Google Suite'))
    else:
        return(row)

In [233]:
df['herramientas'] = df['herramientas'].map(replace_google)

In [235]:
df['herramientas'].value_counts().keys()

Index(['GitHub;GitLab', 'GitHub;Slack;GitLab;Google Suite;Trello',
       'GitHub;Slack;GitLab;Jira;Google Suite;Confluence',
       'GitHub;Slack;GitLab;Jira;Google Suite;Confluence;Microsoft Team;Trello',
       'GitHub;Slack;GitLab;Google Suite;Trello;Discord',
       'GitHub;GitLab;Google Suite;Trello',
       'GitHub;Slack;GitLab;Jira;Google Suite;Trello',
       'GitHub;Slack;GitLab;Jira;Google Suite;Microsoft Team;Trello',
       'GitHub;GitLab;Google Suite', 'Microsoft Team',
       ...
       'GitHub;Slack;Jira;Confluence', 'GitLab;Microsoft Azure',
       'GitHub;Slack;Microsoft Team',
       'Slack;Google Suite;Microsoft Team;Trello',
       'GitHub;GitLab;Jira;Google Suite;Microsoft Team;Discord',
       'GitHub;Slack;Jira;Google Suite;Trello',
       'GitHub;GitLab;Jira;Confluence', 'GitHub;Slack;GitLab;Jira;Discord',
       'Google Suite;Discord',
       'GitHub;Slack;Jira;Google Suite;Confluence;Trello'],
      dtype='object', length=164)

In [239]:
set(itertools.chain.from_iterable([i.split(';') for i in df['herramientas'].value_counts().keys()]))

{'Confluence',
 'Discord',
 'Facebook Workplace',
 'GitHub',
 'GitLab',
 'Google Suite',
 'Jira',
 'Microsoft Azure',
 'Microsoft Team',
 'Slack',
 'Trello'}

In [240]:
# Exporting the df.
df.to_csv('enc_devs_py_2021_fixed.csv', index=False)

In [241]:
# Whit this changes now it is posible to work with the data.